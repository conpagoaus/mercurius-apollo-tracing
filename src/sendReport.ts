import { IReport, Report } from 'apollo-reporting-protobuf'
import { request } from 'undici'

export const sendReport = async (
  report: IReport,
  options: { endpointUrl: string; apiKey: string }
) => {
  return request(
    `${
      options.endpointUrl || 'https://usage-reporting.api.apollographql.com'
    }/api/ingress/traces`,
    {
      method: 'POST',
      headers: {
        'user-agent': 'ApolloServerPluginUsageReporting',
        'x-api-key': options.apiKey,
        'content-encoding': 'gzip',
        accept: 'application/json'
      },
      body: Report.encode(report).finish()
    }
  )
}
