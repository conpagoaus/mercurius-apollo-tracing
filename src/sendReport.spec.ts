import http from 'http'
import zlib from 'zlib'

import { Report } from 'apollo-reporting-protobuf'
import tap from 'tap'
import {
  MockAgent,
  setGlobalDispatcher,
  getGlobalDispatcher,
  Dispatcher
} from 'undici'
import sinon from 'sinon'

import { sendReport } from './sendReport'

const fakeReport = {
  header: {
    hostname: 'www.example.com',
    graphRef: 'alskncka384u1923e8uino1289jncvo019n'
  },
  tracesPerQuery: {
    '# Foo\nquery Foo { user { email pets { name } } }': {
      trace: [],
      statsWithContext: [],
      internalTracesContributingToStats: []
    }
  }
}

tap.test('sendReport encodes the report', async (t) => {
  const requestListener = function (req, res) {
    const chunks: any[] = []
    req.on('data', (chunk: any) => chunks.push(chunk))
    req.on('end', () => {
      const data = Buffer.concat(chunks)

      const unzippedData = zlib.unzipSync(data)

      const reportDecoded = Report.decode(unzippedData)
      t.same(reportDecoded, fakeReport)
    })
    res.writeHead(200)
    res.end()
  }

  const server = http.createServer(requestListener)
  server.listen(3334)

  const res = await sendReport(
    fakeReport,
    {
      endpointUrl: 'http://localhost:3334',
      apiKey: 'fakeKey',
      graphRef: 'myGraph@current'
    },
    {} as any
  )

  t.equal(res.statusCode, 200)

  server.close()
  t.end()
})

tap.test('with mocked http', async (t) => {
  const endpointUrl = 'http://www.example.com'

  let agent: MockAgent
  let originalDispatcher: Dispatcher
  t.beforeEach(() => {
    agent = new MockAgent()
    agent.disableNetConnect()
    originalDispatcher = getGlobalDispatcher()
    setGlobalDispatcher(agent)
  })
  t.afterEach(() => {
    agent.enableNetConnect()
    setGlobalDispatcher(originalDispatcher)
  })
  t.test('thing', async (tt) => {
    return tt.equal(2, 2)
  })
  t.test('sendReport success', async (tt) => {
    agent
      .get(endpointUrl)
      .intercept({
        path: '/api/ingress/traces',
        method: 'POST'
      })
      .reply(200, {})

    const res = await sendReport(
      fakeReport,
      {
        endpointUrl,
        apiKey: 'fakeKey',
        graphRef: 'myGraph@current'
      },
      {} as any
    )
    return tt.equal(res.statusCode, 200)
  })
  t.test('sendReport error', async (tt) => {
    agent
      .get(endpointUrl)
      .intercept({
        path: '/api/ingress/traces',
        method: 'POST'
      })
      .reply(400, 'something went wrong')

    const logErrorSpy = sinon.spy()
    const res = await sendReport(
      fakeReport,
      {
        endpointUrl,
        apiKey: 'fakeKey',
        graphRef: 'myGraph@current'
      },
      {
        log: {
          error: logErrorSpy
        }
      } as any
    )
    tt.equal(res.statusCode, 400)
    return tt.equal(logErrorSpy.called, true)
  })
})
