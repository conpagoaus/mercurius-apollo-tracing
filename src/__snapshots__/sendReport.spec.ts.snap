// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendReport should send 1`] = `
Object {
  "header": Object {
    "graphRef": "alskncka384u1923e8uino1289jncvo019n",
    "hostname": "www.example.com",
  },
  "tracesPerQuery": Object {
    "# Foo
query Foo { user { email pets { name } } }": Object {
      "internalTracesContributingToStats": Array [],
      "statsWithContext": Array [],
      "trace": Array [],
    },
  },
}
`;
